osbuild/util/path.py

```diff
@@ -1,5 +1,39 @@
 """Path handling utility functions"""
+import errno
 import os
+import os.path
+
+from .ctx import suppress_oserror
+
+
+def clamp_mtime(path: str, start: int, to: int):
+    """Clamp all modification times of 'path'
+
+    Set the mtime of 'path' to 'to' if it is greater or equal to 'start'.
+    If 'to' is None, the mtime is set to the current time.
+    """
+
+    times = (to, to)
+
+    def fix_utime(path, dfd: int):
+        sb = os.stat(path, dir_fd=dfd, follow_symlinks=False)
+        if sb.st_mtime < start:
+            return
+
+        # We might get a permission error when the immutable flag is set;
+        # since there is nothing much we can do, we just ignore it
+        with suppress_oserror(errno.EPERM):
+            os.utime(path, times, dir_fd=dfd, follow_symlinks=False)
+
+    for _, _, files, dfd in os.fwalk(path):
+
+        # Clamp the mtime of the current directory, this also ensures
+        # that all sub-directores are taken care of since `os.fwalk`
+        # will also be called on each of them.
+        fix_utime(".", dfd)
+
+        for f in files:
+            fix_utime(f, dfd)
```

Comment:
I think this loop skips symlinks. Maybe you can add a test to verify this?

If I read `fwalk()` (and `walk()`) correctly, they follow a symlink to determine whether it points to a directory or a file or is dangling. If it points to a directory, it is not included in `files`. Only file-pointers or dangling symlinks are included in `files`.

This behavior cannot be changed, even if `follow_symlinks` is `False` (which is the default, anyway).

https://github.com/python/cpython/blob/3.11/Lib/os.py#L479

I cannot understand that decision and I think it is completely wrong. But that's how it is. I think the correct thing to do is walk the union of `dirs` and `files` and check manually whether it is a symlink first, and then check whether it is a dir afterwards.

Alternatively, in this case here, just apply `fix_utime()` unconditionally to both dirs and files?